pkg/controller/core/leader_aware_reconciler.go

```diff
@@ -0,0 +1,86 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	config "sigs.k8s.io/kueue/apis/config/v1beta1"
+)
+
+// WithLeadingManager returns a decorating reconcile.Reconciler that discards reconciliation requests
+// for the controllers that are started with the controller.Options.NeedLeaderElection
+// option set to false in non-leading replicas.
+//
+// Starting controllers in non-leading replicas is needed for these that update the data
+// served by the visibility extension API server.
+//
+// This enables to:
+//   - Keep the scheduling decisions under the responsibility of the leading replica alone,
+//     to prevent any concurrency issues.
+//   - Consume requests from the watch event queues, to prevent them from growing indefinitely
+//     in the non-leading replicas.
+//   - Transition to actually reconciling requests in the replica that may acquire
+//     the leader election lease, in case the previously leading replica failed to renew it.
+func WithLeadingManager(mgr ctrl.Manager, reconciler reconcile.Reconciler, cfg *config.Configuration) reconcile.Reconciler {
+	// Do not decorate the reconciler if leader election is disabled
+	if cfg.LeaderElection == nil || !ptr.Deref(cfg.LeaderElection.LeaderElect, false) {
+		return reconciler
+	}
+
+	// Default to the recommended lease duration, that's used for core components
+	requeueDuration := 15 * time.Second
+	// Otherwise used the configured lease duration for the manager
+	zero := metav1.Duration{}
+	if duration := cfg.LeaderElection.LeaseDuration; duration != zero {
+		requeueDuration = duration.Duration
+	}
+
+	return &leaderAwareReconciler{
+		elected:         mgr.Elected(),
+		delegate:        reconciler,
+		requeueDuration: requeueDuration,
+	}
+}
+
+type leaderAwareReconciler struct {
+	elected         <-chan struct{}
+	delegate        reconcile.Reconciler
+	requeueDuration time.Duration
+}
+
+var _ reconcile.Reconciler = (*leaderAwareReconciler)(nil)
+
+func (r *leaderAwareReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	select {
+	case <-r.elected:
+		// The manager has been elected leader, delegate reconciliation to the provided reconciler.
+		return r.delegate.Reconcile(ctx, request)
+	default:
+		// The manager hasn't been elected leader yet, requeue the reconciliation request
+		// to prevent against any missed / discarded events over the period it takes
+		// to fail over a new leading replica, which can take as much as the configured
+		// lease duration, for it to acquire leadership.
+		return ctrl.Result{RequeueAfter: r.requeueDuration}, nil
```

Comment:
If a replica doesn't get the lease for a long time, we will keep requeueing keys for objects that might have been deleted. This would be a form of memory leak. Could we quickly check if the objects exist in the client?

Reply:
I've added the logic that prevent from re-enqueuing requests indefinitely if the object has been deleted / is not found. PTAL.